finetuner/tuner/dataset/samplers.py

```diff
@@ -0,0 +1,220 @@
+from collections import defaultdict
+from copy import deepcopy
+from math import ceil
+from random import choices, sample, shuffle
+from typing import Iterator, Sequence, Tuple
+
+
+class RandomClassBatchSampler:
+    """A batch sampler that fills the batch with an equal number of items from each
+    class.
+
+    It will try to make sure that all the items get used once in a single epoch, and
+    that only ``num_items_per_class`` items in a batch come from one class. When there
+    would not be enough items left from a single class to fill the batch, items will be
+    randomly sampled from other already used items for this class.
+
+    However, some cutoff might occur if there are not enough items left from different
+    classes to fill the batch - in this case some items do not get used in that epoch.
+    """
+
+    def __init__(
+        self, labels: Sequence[int], batch_size: int, num_items_per_class: int
+    ):
+        """Construct the batch sample.
+
+        :param labels: A sequence of items labels, each label should be an integer
+            denoting the class of the item
+        :param batch_size: How many items to include in a batch
+        :param num_items_per_class: How many items per class (unique labels) to include
+            in a batch. For example, if ``batch_size`` is 20, and
+            ``num_items_per_class`` is 4, the batch will consist of 4 items for each of
+            the 5 classes.
+        """
+        if batch_size <= 0:
+            raise ValueError('batch_size must be a positive integer')
+        if num_items_per_class <= 0:
+            raise ValueError('num_items_per_class must be a positive integer')
+
+        if batch_size % num_items_per_class != 0:
+            raise ValueError(
+                'batch_size must be a multiple of num_items_per_class,'
+                f' got batch_size={batch_size} and'
+                f' num_items_per_class={num_items_per_class}'
+            )
+
+        self._num_items_per_class = num_items_per_class
+        self._batch_size = batch_size
+        self._num_classes = batch_size // num_items_per_class
+
+        if len(set(labels)) < self._num_classes:
+            raise ValueError(
+                'Number of distinct classes in labels must be larger or equal than'
+                'numbers of classes in a batch'
+            )
+
```

Comment:
the logic here is too complex, can we assume unique `labels` we passed into the class is already denotes `num_classes`? then `num_items_per_class` can be computed by `batch_size % len(set(labels))`

Reply:
No, `num_classes` and `num_items_per_class` refer to a single batch - if you have a dataset with 10000 unique labels you obviously can not fit them all into a single batch of length 64.

Reply:
okay